sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/QueryPlan.scala

```diff
@@ -67,15 +67,16 @@ abstract class QueryPlan[PlanType <: QueryPlan[PlanType]]
   def outputOrdering: Seq[SortOrder] = Nil
 
   // Override `treePatternBits` to propagate bits for its expressions.
-  override lazy val treePatternBits: BitSet = {
+  private val _treePatternBits = new BestEffortLazyVal[BitSet](() => {
```

Comment:
`treePatternBits` was kept `lazy val` in [previous PR](https://github.com/apache/spark/pull/49212), do you @cloud-fan remember it was intended or not?

Reply:
Curiously, Since the `treePatternBits` only init once, how it keep the correctness if the expressions transformed with rules?

Reply:
> if the expressions transformed with rules?

Rules will return new expressions? Catalyst operators/expressions are immutable, from rules' perspective.